GeoLib/DuplicateGeometry.cpp

```diff
@@ -0,0 +1,121 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#include "DuplicateGeometry.h"
+
+// ThirdParty/logog
+#include <logog/include/logog.hpp>
+
+#include "GeoLib/GEOObjects.h"
+#include "GeoLib/Point.h"
+#include "GeoLib/Polyline.h"
+#include "GeoLib/Surface.h"
+#include "GeoLib/Triangle.h"
+
+
+namespace GeoLib
+{
+
+DuplicateGeometry::DuplicateGeometry(GeoLib::GEOObjects &geo_objects,
+    std::string const& input_name,
+    std::string const& output_name)
+: _output_name(output_name), _geo_objects(geo_objects)
+{
+    duplicate(input_name);
+}
+
+void DuplicateGeometry::duplicate(std::string const& input_name)
+{
+    std::vector<GeoLib::Point*> const*const pnts (_geo_objects.getPointVec(input_name));
+    if (pnts == nullptr)
+    {
+        ERR("Geometry \"%s\" not found.", input_name.c_str());
+        return;
+    }
+
+    std::unique_ptr< std::vector<GeoLib::Point*> > new_pnts (new std::vector<GeoLib::Point*>);
+    new_pnts->reserve(pnts->size());
+    std::transform(pnts->cbegin(), pnts->cend(), std::back_inserter(*new_pnts),
+        [](GeoLib::Point* point) { return new GeoLib::Point(*point); });
+    _geo_objects.addPointVec(std::move(new_pnts), _output_name);
+
+    std::vector<GeoLib::Polyline*> const* plys (_geo_objects.getPolylineVec(input_name));
+    if (plys)
+    {
+        auto new_plys = copyPolylinesVector(*plys);
+        _geo_objects.addPolylineVec(std::move(new_plys), _output_name);
+    }
+
+    std::vector<GeoLib::Surface*> const* sfcs (_geo_objects.getSurfaceVec(input_name));
+    if (sfcs)
+    {
+        auto new_sfcs = copySurfacesVector(*sfcs);
+        _geo_objects.addSurfaceVec(std::move(new_sfcs), _output_name);
+    }
+}
+
+std::unique_ptr<std::vector<GeoLib::Polyline*>> DuplicateGeometry::copyPolylinesVector(
+    std::vector<GeoLib::Polyline*> const& polylines)
```

Comment:
Not sure here, but if the method is not modifying anything it could be declared const.